core/src/main/java/com/linecorp/armeria/internal/client/dns/CachingDnsResolver.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.client.dns;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.DnsCache;
+import com.linecorp.armeria.common.util.AbstractUnwrappable;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsRecord;
+
+final class CachingDnsResolver extends AbstractUnwrappable<DnsResolver> implements DnsResolver {
+
+    private static final Logger logger = LoggerFactory.getLogger(CachingDnsResolver.class);
+
+    private final Map<DnsQuestion, CompletableFuture<List<DnsRecord>>> inflightRequests =
```

Comment:
I intentionally didn't use `AsyncCache` because:
- `inflightRequests` is temporary storage so it does not cache something. `DnsCache` manages the cached data.
- `AsyncCache` built on the top of `ConcurrentHashMap` provides various features such as metrics, listeners that might be more expensive than `ConcurrentHashMap`.